File: redistributable/openstack.net/src/Samples/CPPCodeSamples/QueueingServiceExamples.cpp
#include "Stdafx.h"

using namespace net::openstack::Core;
using namespace net::openstack::Core::Collections;
using namespace net::openstack::Core::Domain;
using namespace net::openstack::Core::Providers;
using namespace net::openstack::Core::Domain::Queues;
using namespace net::openstack::Providers::Rackspace;
using namespace System;
using namespace System::Collections::ObjectModel;
using namespace System::Threading;
using namespace System::Threading::Tasks;

ref class QueueingServiceExamples
{
private:
	static CloudIdentity^ identity = gcnew CloudIdentity();
	static String^ region = nullptr;
	static Guid clientId = Guid::NewGuid();
	static bool internalUrl = false;
	static IIdentityProvider^ identityProvider = nullptr;

public:
	void GetHome()
	{
#pragma region GetHomeAsync (TPL)
		IQueueingService^ queueingService = gcnew CloudQueuesProvider(identity, region, clientId, internalUrl, identityProvider);
		Task<HomeDocument^>^ task = queueingService->GetHomeAsync(CancellationToken::None);
#pragma endregion
	}

	void GetNodeHealth()
	{
#pragma region GetNodeHealthAsync (TPL)
		IQueueingService^ queueingService = gcnew CloudQueuesProvider(identity, region, clientId, internalUrl, identityProvider);
		Task^ task = queueingService->GetNodeHealthAsync(CancellationToken::None);
#pragma endregion
	}

	void CreateQueue()
	{
#pragma region CreateQueueAsync (TPL)
		IQueueingService^ queueingService = gcnew CloudQueuesProvider(identity, region, clientId, internalUrl, identityProvider);
		QueueName^ queueName = gcnew QueueName("ExampleQueue");
		Task<bool>^ task = queueingService->CreateQueueAsync(queueName, CancellationToken::None);
#pragma endregion
	}

	void DeleteQueue()
	{
#pragma region DeleteQueueAsync (TPL)
		IQueueingService^ queueingService = gcnew CloudQueuesProvider(identity, region, clientId, internalUrl, identityProvider);
		QueueName^ queueName = gcnew QueueName("ExampleQueue");
		Task^ task = queueingService->DeleteQueueAsync(queueName, CancellationToken::None);
#pragma endregion
	}

#pragma region ListQueuesAsync (TPL)
	void ListQueues()
	{
		IQueueingService^ queueingService = gcnew CloudQueuesProvider(identity, region, clientId, internalUrl, identityProvider);
		Task<ReadOnlyCollectionPage<CloudQueue^>^>^ queuesPageTask = queueingService->ListQueuesAsync(nullptr, Nullable<int>(), true, CancellationToken::None);
		auto func = gcnew Func<Task<ReadOnlyCollectionPage<CloudQueue^>^>^, Task<ReadOnlyCollection<CloudQueue^>^>^>(GetAllPagesContinuationAsync<CloudQueue^>);
		Task<ReadOnlyCollection<CloudQueue^>^>^ queuesTask = TaskExtensions::Unwrap(queuesPageTask->ContinueWith(func));
	}

	generic<class T>
	static Task<ReadOnlyCollection<T>^>^ GetAllPagesContinuationAsync(Task<ReadOnlyCollectionPage<T>^>^ pageTask)
	{
		return ReadOnlyCollectionPageExtensions::GetAllPagesAsync(pageTask->Result, CancellationToken::None, static_cast<IProgress<ReadOnlyCollectionPage<T>^>^>(nullptr));
	}
#pragma endregion

	void QueueExists()
	{
#pragma region QueueExistsAsync (TPL)
		IQueueingService^ queueingService = gcnew CloudQueuesProvider(identity, region, clientId, internalUrl, identityProvider);
		QueueName^ queueName = gcnew QueueName("ExampleQueue");
		Task<bool>^ task = queueingService->QueueExistsAsync(queueName, CancellationToken::None);
#pragma endregion
	}
};

File: redistributable/openstack.net/src/Samples/FSharpCodeSamples/QueueingServiceExamples.fs
module QueueingServiceExamples

open System
open System.Threading
open System.Threading.Tasks
open net.openstack.Core
open net.openstack.Core.Collections
open net.openstack.Core.Domain
open net.openstack.Core.Domain.Queues
open net.openstack.Core.Providers
open net.openstack.Providers.Rackspace

let identity = new CloudIdentity (Username = "MyUser", APIKey = "API_KEY_HERE")
let region : string = null
let clientId = Guid.NewGuid()
let internalUrl = false
let identityProvider : IIdentityProvider = null

let getHomeAsyncAwait =
    async {
        //#region GetHomeAsync (await)
        let queueingService = new CloudQueuesProvider(identity, region, clientId, internalUrl, identityProvider)
        let! homeDocument = queueingService.GetHomeAsync(CancellationToken.None) |> Async.AwaitTask
        //#endregion
        ()
    }

let getHome =
    //#region GetHomeAsync (TPL)
    let queueingService = new CloudQueuesProvider(identity, region, clientId, internalUrl, identityProvider)
    let task = queueingService.GetHomeAsync(CancellationToken.None)
    //#endregion
    ()

let getNodeHealthAsyncAwait =
    async {
        //#region GetNodeHealthAsync (await)
        let queueingService = new CloudQueuesProvider(identity, region, clientId, internalUrl, identityProvider)
        queueingService.GetNodeHealthAsync(CancellationToken.None) |> Async.AwaitIAsyncResult |> ignore
        //#endregion
        ()
    }

let getNodeHealth =
    //#region GetNodeHealthAsync (TPL)
    let queueingService = new CloudQueuesProvider(identity, region, clientId, internalUrl, identityProvider)
    let task = queueingService.GetNodeHealthAsync(CancellationToken.None)
    //#endregion
    ()

let createQueueAsyncAwait =
    async {
        //#region CreateQueueAsync (await)
        let queueingService = new CloudQueuesProvider(identity, region, clientId, internalUrl, identityProvider)
        let queueName = new QueueName("ExampleQueue")
        let! createdQueue = queueingService.CreateQueueAsync(queueName, CancellationToken.None) |> Async.AwaitTask
        //#endregion
        ()
    }

let createQueue =
    //#region CreateQueueAsync (TPL)
    let queueingService = new CloudQueuesProvider(identity, region, clientId, internalUrl, identityProvider)
    let queueName = new QueueName("ExampleQueue")
    let task = queueingService.CreateQueueAsync(queueName, CancellationToken.None)
    //#endregion
    ()

let deleteQueueAsyncAwait =
    async {
        //#region DeleteQueueAsync (await)
        let queueingService = new CloudQueuesProvider(identity, region, clientId, internalUrl, identityProvider)
        let queueName = new QueueName("ExampleQueue")
        queueingService.DeleteQueueAsync(queueName, CancellationToken.None) |> Async.AwaitIAsyncResult |> ignore
        //#endregion
        ()
    }

let deleteQueue =
    //#region DeleteQueueAsync (TPL)
    let queueingService = new CloudQueuesProvider(identity, region, clientId, internalUrl, identityProvider)
    let queueName = new QueueName("ExampleQueue")
    let task = queueingService.DeleteQueueAsync(queueName, CancellationToken.None)
    //#endregion
    ()

let listQueuesAsyncAwait =
    async {
        //#region ListQueuesAsync (await)
        let queueingService = new CloudQueuesProvider(identity, region, clientId, internalUrl, identityProvider)
        let! queuesPage = queueingService.ListQueuesAsync(null, Nullable(), true, CancellationToken.None) |> Async.AwaitTask
        let! queues = queuesPage.GetAllPagesAsync(CancellationToken.None, null) |> Async.AwaitTask
        //#endregion
        ()
    }

let listQueues =
    //#region ListQueuesAsync (TPL)
    let queueingService = new CloudQueuesProvider(identity, region, clientId, internalUrl, identityProvider)
    let queuesPageTask = queueingService.ListQueuesAsync(null, Nullable(), true, CancellationToken.None)
    let queuesTask = queuesPageTask.ContinueWith(fun (task:Task<ReadOnlyCollectionPage<CloudQueue>>) -> task.Result.GetAllPagesAsync(CancellationToken.None, null)) |> TaskExtensions.Unwrap
    //#endregion
    ()

let queueExistsAsyncAwait =
    async {
        //#region QueueExistsAsync (await)
        let queueingService = new CloudQueuesProvider(identity, region, clientId, internalUrl, identityProvider)
        let queueName = new QueueName("ExampleQueue")
        let! exists = queueingService.QueueExistsAsync(queueName, CancellationToken.None) |> Async.AwaitTask
        //#endregion
        ()
    }

let queueExists =
    //#region QueueExistsAsync (TPL)
    let queueingService = new CloudQueuesProvider(identity, region, clientId, internalUrl, identityProvider)
    let queueName = new QueueName("ExampleQueue")
    let task = queueingService.QueueExistsAsync(queueName, CancellationToken.None)
    //#endregion
    ()
